[Gutenberg Data Tutorial] Part 4: Building a new record form (#39261)
WordPress/gutenberg: docs/how-to-guides/data-basics/4-building-a-create-page-form.md

# Part 4: Building a _Create page_ form

In the [previous part](/docs/how-to-guides/data-basics/3-building-an-edit-form.md) we created an *Edit page* feature, and in this part we will add a *Create page* feature. Here's a glimpse of what we're going to build:

![](/docs/how-to-guides/data-basics/media/create-form/create-form-with-text.png)

### Step 1: Add a _Create a new page_ button

Let’s start by building a button to display the _create page_ form. It’s similar to an _Edit_ button we have built in the [part 3](/docs/how-to-guides/data-basics/3-building-an-edit-form.md):

```js
import { useDispatch } from '@wordpress/data';
import { Button, Modal, TextControl } from '@wordpress/components';

> function PageCreateButton() {

Reviewer: Can this be named `CreatePageButton` ?

	const [ isOpen, setOpen ] = useState( false );
	const openModal = () => setOpen( true );
	const closeModal = () => setOpen( false );
	return (
		<>
			<Button onClick={ openModal } variant="primary">

> 				Create a new Page

Reviewer 2: I know we don't need translation functions here but is it better to use them to show the best practice?

Author: Let's add that to all parts in a follow-up PR to avoid suddenly introducing translations in part 4.

			</Button>
			{ isOpen && (
				<Modal onRequestClose={ closeModal } title="Create a new page">
					<CreatePageForm
						onCancel={ closeModal }
						onSaveFinished={ closeModal }
					/>
				</Modal>
			) }
		</>
	);
}

function CreatePageForm() {
	// Empty for now
	return <div />;
}

```

Great! Now let’s make `MyFirstApp` display our shiny new button:

```js
function MyFirstApp() {
	// ...
	return (
		<div>
			<div className="list-controls">
				<SearchControl onChange={ setSearchTerm } value={ searchTerm } />
				<PageCreateButton />
			</div>
			<PagesList hasResolved={ hasResolved } pages={ pages } />
		</div>
	);
}
```

The final result should look as follows:

![](/docs/how-to-guides/data-basics/media/create-form/create-button.png)

### Step 2: Extract a controlled PageForm

> Now that the button is in place, we may focus entirely on building the form. This tutorial is about managing data, so we will not build a complete page editor. Instead, the form will only contain one field: post title.

Reviewer 3: ```suggestion
Now that the button is in place, we can focus entirely on building the form. This tutorial is about managing data, so we will not build a complete page editor. Instead, the form will only contain one field: post title.
```

Nit: tiny flow thing.

> Luckily, the `EditPageForm` we have built in [part three](/docs/how-to-guides/data-basics/3-building-an-edit-form.md) already takes us 80% there. The bulk of the user interface is already there, and we will reuse it in the `CreatePageForm`. Let’s start by extracting the form UI into a separate component:

Reviewer 3: ```suggestion
Luckily, the `EditPageForm` we built in [part three](/docs/how-to-guides/data-basics/3-building-an-edit-form.md) already takes us 80% of the way there. The bulk of the user interface is already available, and we will reuse it in the `CreatePageForm`. Let’s start by extracting the form UI into a separate component:
```

Nit: tiny flow things.

```js
export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
	// ...
	return (
		<PageForm
			title={ page.title }
			onChangeTitle={ handleChange }
			hasEdits={ hasEdits }
			lastError={ lastError }
			isSaving={ isSaving }
			onCancel={ onCancel }
			onSave={ handleSave }
		/>
	);
}

export function PageForm( { title, onChangeTitle, hasEdits, lastError, isSaving, onCancel, onSave } ) {
	return (
		<div className="my-gutenberg-form">
			<TextControl
				label="Page title:"
				value={ title }
				onChange={ onChangeTitle }
			/>
			{ lastError ? (
				<div className="form-error">Error: { lastError.message }</div>
			) : (
				false
			) }
			<div className="form-buttons">
				<Button
					onClick={ onSave }
					variant="primary"
					disabled={ ! hasEdits || isSaving }
				>

> 					{ isSaving ? <Spinner /> : 'Save' }

Reviewer 3: a11y - if you have a button with text and that suddenly changes to a spinner there is nothing to communicate what is happening. Possibly needs a label to say `Saving...`.

				</Button>

> 				<Button onClick={ onCancel } variant="tertiary">

Reviewer 3: What happens if I hit cancel when it's saving?

Author: Good question! There are no special semantics, you'll just close the modal: 

```js
<EditPageForm
	pageId={ pageId }
	onCancel={ closeModal }
	onSaveFinished={ closeModal }
/>
```

Maybe this button could be disabled while saving to make it explicit that you cannot really cancel the save action.

Author: I just clarified that in the part 3 and carried over the solution to this part.

					Cancel
				</Button>
			</div>
		</div>
	);
}
```

> This code quality change should not change anything about how the application works. Let’s try to edit a page just to be sure:

Reviewer 3: ```suggestion
This code quality change should not alter anything about how the application works. Let’s try to edit a page just to be sure:
```

Nit: avoid using same word twice?

![](/docs/how-to-guides/data-basics/media/create-form/edit-page-form.png)

Great! The edit form is still there, and now we have a building block to power the new `CreatePageForm`.

### Step 3: Build a CreatePageForm

The only thing that `CreatePageForm` component must do is to provide the following seven properties needed to render the `PageForm` component:

* title
* onChangeTitle
* hasEdits
* lastError
* isSaving
* onCancel
* onSave

Let’s see how we can do that:

#### Title, onChangeTitle, hasEdits

The `EditPageForm` updated and saved an existing entity record that lived in the Redux state. Because of that, we relied on the `editedEntityRecords` selector.

> In case of the `CreatePageForm`, however, there is no pre-existing entity record. There is only an empty form. Anything that the user types is local to that form, which means we can kep track of it using the React’s `useState` hook:

Reviewer 3: ```suggestion
In case of the `CreatePageForm` however, there is no pre-existing entity record. There is only an empty form. Anything that the user types is local to that form, which means we can keep track of it using the React’s `useState` hook:
```

```js
export function CreatePageForm( { onCancel, onSaveFinished } ) {
	const [ title, setTitle ] = useState();
	const handleChange = ( title ) => setTitle( title );
	return (
		<PageForm
			title={ title }
			onChangeTitle={ setTitle }
			hasEdits={ !! title }
			{ /* ... */ }
			/>
	);
}
```

#### onSave, onCancel

In the `EditPageForm`, we dispatched the `saveEditedEntityRecord('postType', 'page', pageId )` action to save the edits that lived in the Redux state.

> In the `CreatePageForm`, however, we do not have any edits in the Redux state, nor we do have any `pageId`. The action we need to dispatch in this case is called [`saveEntityRecord`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#saveentityrecord) (without the word _Edited_ in the name) and it accepts an object representing the new entity record instead of a `pageId`.

Reviewer 3: ```suggestion
In the `CreatePageForm` however, we do not have any edits in the Redux state, nor we do have a `pageId`. The action we need to dispatch in this case is called [`saveEntityRecord`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#saveentityrecord) (without the word _Edited_ in the name) and it accepts an object representing the new entity record instead of a `pageId`.
```

The data passed to `saveEntityRecord` is sent via a POST request to the appropriate REST API endpoint. For example, dispatching the following action:

```js
saveEntityRecord( 'postType', 'page', { title: "Test" } );
```

Triggers a POST request to the [`/wp/v2/pages` WordPress REST API](https://developer.wordpress.org/rest-api/reference/pages/) endpoint with a single field in the request body: `title=Test`.

Now that we know more about `saveEntityRecord`, let's use it in `CreatePageForm`.

```js
export function CreatePageForm( { onSaveFinished, onCancel } ) {
	// ...
	const { saveEntityRecord } = useDispatch( coreDataStore );
	const handleSave = async () => {
		const savedRecord = await saveEntityRecord(
			'postType',
			'page',
			{ title }
		);
		if ( savedRecord ) {
			onSaveFinished();
		}
	};
	return (
		<PageForm
			{ /* ... */ }
			onSave={ handleSave }
			onCancel={ onCancel }
			/>
	);
}
```

> There is one more detail to address: our newly created pages are not yet picked up by the `PagesList`. Accordingly to the REST API documentation,  the `/wp/v2/pages` endpoint creates pages with `status=draft` by default, but returns pages with `status=publish`. The solution is to pass the `status` parameter explicitly:

Reviewer 3: ```suggestion
There is one more detail to address: our newly created pages are not yet picked up by the `PagesList`. Accordingly to the REST API documentation,  the `/wp/v2/pages` endpoint creates (`POST` requests) pages with `status=draft` by default, but _returns_ (`GET` requests) pages with `status=publish`. The solution is to pass the `status` parameter explicitly:
```

```js
export function CreatePageForm( { onSaveFinished, onCancel } ) {
	// ...
	const { saveEntityRecord } = useDispatch( coreDataStore );
	const handleSave = async () => {
		const savedRecord = await saveEntityRecord(
			'postType',
			'page',
			{ title, status: 'publish' }
		);
		if ( savedRecord ) {
			onSaveFinished();
		}
	};
	return (
		<PageForm
			{ /* ... */ }
			onSave={ handleSave }
			onCancel={ onCancel }
			/>
	);
}
```

Go ahead and apply that change to your local `CreatePageForm` component, and let’s tackle the remaining two props.

#### lastError, isSaving

The `EditPageForm`  retrieved the error and progress information via the `getLastEntitySaveError`  and `isSavingEntityRecord` selectors. In both cases, it passed the following three arguments: `( 'postType', 'page', pageId )`.

> In `CreatePageForm`, however, we do not have a `pageId`. What now? We skip the `pageId` argument to retrieve the information about the entity record without any id – which is the newly created one. The `useSelect` call is thus very similar to the one from `EditPageForm`:

Reviewer 3: ```suggestion
In `CreatePageForm` however, we do not have a `pageId`. What now? We can skip the `pageId` argument to retrieve the information about the entity record without any id – this will be the newly created one. The `useSelect` call is thus very similar to the one from `EditPageForm`:
```
